v2/internal/reconcilers/mysql/mysql_user_reconciler.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return result, nil
+}
+
+func (r *MySQLUserReconciler) Delete(ctx context.Context, log logr.Logger, user *asomysql.User) (ctrl.Result, error) {
+	// TODO: A lot of this is duplicated. See https://github.com/Azure/azure-service-operator/pull/2327 for a proposal to fix that
+	log.V(Status).Info("Starting delete of resource")
+
+	// If we have no resourceID to begin with, or no finalizer, the Azure resource was never created
+	hasFinalizer := controllerutil.ContainsFinalizer(user, reconcilers.GenericControllerFinalizer)
+	if !hasFinalizer {
+		return ctrl.Result{}, r.RemoveResourceFinalizer(ctx, log, user)
+	}
+
+	reconcilePolicy := reconcilers.GetReconcilePolicy(user, log)
+	if !reconcilePolicy.AllowsDelete() {
+		log.V(Info).Info("Bypassing delete of resource due to policy", "policy", reconcilePolicy)
+		return ctrl.Result{}, r.RemoveResourceFinalizer(ctx, log, user)
+	}
+
+	// Check that this objects owner still exists
+	// This is an optimization to avoid excess requests to Azure.
+	_, err := r.ResourceResolver.ResolveOwner(ctx, user)
+	if err != nil {
+		var typedErr *resolver.ReferenceNotFound
+		if errors.As(err, &typedErr) {
+			return ctrl.Result{}, r.RemoveResourceFinalizer(ctx, log, user)
+		}
+	}
+
+	secrets, err := r.ResourceResolver.ResolveResourceSecretReferences(ctx, user)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	db, err := r.connectToDB(ctx, log, user, secrets)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	defer db.Close()
+
+	// TODO: There's still probably some ways that this user can be deleted but that we don't detect (and
+	// TODO: so might cause an error triggering the resource to get stuck).
+	// TODO: We check for owner not existing above, but cases where the server is in the process of being
+	// TODO: deleted (or all system tables have been wiped?) might also exist...
+	err = mysqlutil.DropUser(ctx, db, user.Spec.AzureName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.RemoveResourceFinalizer(ctx, log, user)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MySQLUserReconciler) CreateOrUpdate(ctx context.Context, log logr.Logger, user *asomysql.User) (ctrl.Result, error) {
+	if r.NeedToClaimResource(user) {
+		// TODO: Make this synchronous for both ARM resources and here
+		// TODO: Will do this in a follow-up PR
+		_, err := r.ClaimResource(ctx, log, user)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Resolve the secrets
+	secrets, err := r.ResourceResolver.ResolveResourceSecretReferences(ctx, user)
+	if err != nil {
+		return ctrl.Result{}, reconcilers.ClassifyResolverError(err)
+	}
+
+	db, err := r.connectToDB(ctx, log, user, secrets)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	defer db.Close()
+
+	// TODO: A lot of this is duplicated. See https://github.com/Azure/azure-service-operator/pull/2327 for a proposal to fix that
```

Comment:
This PR is merged - any other cleanup required?

Reply:
The PR was just the design. Let me update this link to the design doc. I haven't actually _done_ the cleanup yet (well, I have but it's in a branch you haven't seen yet and isn't quite ready to merge yet).

Reply:
Pointed this link to https://azure.github.io/azure-service-operator/design/reconcile-interface/ now